coordinator/src/main/scala/filodb.coordinator/TenantIngestionMetering.scala

```diff
@@ -16,6 +17,87 @@ import filodb.coordinator.client.QueryCommands.LogicalPlan2Query
 import filodb.core.DatasetRef
 import filodb.query.{QueryError, QueryResult, TsCardinalities}
 
+object QueryThrottle {
+  // currently just add this diff to the delay if the timeout rate exceeds THRESHOLD
+  val DELAY_DIFF = FiniteDuration(5L, TimeUnit.MINUTES)
+  // number of past query timeouts/non-timeouts to consider
+  val LOOKBACK = 10
+  // {non-timeouts-in-lookback-window} / LOOKBACK < THRESHOLD will adjust the delay
+  val THRESHOLD = 0.85
+}
+
+/**
+ * Throttles the TenantIngestionMetering query rate according to the ratio of timeouts to non-timeouts.
+ *
+ * @param queryDelay the initial delay between each query. This is the duration to be adjusted.
+ */
+class QueryThrottle(queryDelay: FiniteDuration) extends StrictLogging {
```

Comment:
If the general implementation looks good: I'll move this to a new file, parameterize some of the constants, and give it a proper unit test or two.

Reply:
Done